test-suite/shared/src/test/scala/org/scalajs/testsuite/niocharset/BaseCharsetTest.scala

```diff
@@ -30,6 +30,57 @@ class BaseCharsetTest(val charset: Charset) {
   protected def testDecode(in: ByteBuffer)(
       outParts: OutPart[CharBuffer]*): Unit = {
 
+    def testDecodeVsSteppedDecode(malformedAction: CodingErrorAction,
+        unmappableAction: CodingErrorAction) = {
+
+      val in1 = in.duplicate
+      val in2 = in.duplicate
+
+      in2.limit(in2.position)
+
+      val d1 = charset.newDecoder()
+      d1.onMalformedInput(malformedAction)
+      d1.onUnmappableCharacter(unmappableAction)
+
+
+      val r1 = Try(d1.decode(in1))
+      val out2 = CharBuffer.allocate(in.capacity * 2)
```

Comment:
I suggest clearly grouping the things used to compute `r1` on the one hand, and the things used to compute `r2` on the other hand. At the moment they are interleaved for no obvious reason, so it's harder to understand what `r1` and `r2` represent, respectively.

Even better, put everything used to compute `r1` in the `Try { ... }` for `r1`, and respectively for `r2`. AFAICT they're totally independent, so variables should not be needed in both.

Reply:
And now that I've wrapped my head around what you're testing here, I suggest giving more descriptive names to `r1` and `r2`. Maybe `fullDecodeResult` and `steppedDecodeResult`, respectively?

Reply:
I think directDecodeResult may be even better than fullDecodeResult (considering both should be full).

I'll review all names in this method.